crates/optimism/payload/src/tree_validator.rs

```diff
@@ -0,0 +1,203 @@
+        > + Clone
+        + Unpin
+        + 'static
+{
+}
+
+/// Optimism-specific payload validator that uses [`TreePayloadValidator`] for common validation
+/// logic.
+#[derive(Debug)]
+pub struct OpPayloadValidator<P, C, Spec>
+where
+    P: OpProvider,
+    C: ConfigureEvm<Primitives = OpPrimitives> + 'static,
+{
+    /// Reusable validation logic provider
+    tree_validator: TreePayloadValidator<OpPrimitives, P, C>,
+    /// Chain spec for hardfork checks
+    chain_spec: Arc<Spec>,
+}
+
+impl<P, C, Spec> OpPayloadValidator<P, C, Spec>
+where
+    P: OpProvider,
+    C: ConfigureEvm<Primitives = OpPrimitives> + 'static,
+    Spec: OpHardforks,
+{
+    /// Creates a new Optimism payload validator.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        provider: P,
+        consensus: Arc<dyn FullConsensus<OpPrimitives, Error = ConsensusError>>,
+        evm_config: C,
+        config: TreeConfig,
+        payload_processor: PayloadProcessor<OpPrimitives, C>,
+        precompile_cache_map: PrecompileCacheMap<SpecFor<C>>,
+        invalid_block_hook: Box<dyn InvalidBlockHook<OpPrimitives>>,
+        chain_spec: Arc<Spec>,
+    ) -> Self {
+        let tree_validator = TreePayloadValidator::new(
+            provider,
+            consensus,
+            evm_config,
+            config,
+            payload_processor,
+            precompile_cache_map,
+            100, // invalid_headers_cache_size
+            invalid_block_hook,
+            Default::default(), // metrics
+        );
+
+        Self { tree_validator, chain_spec }
+    }
+}
+
+impl<P, C, Spec> PayloadValidator for OpPayloadValidator<P, C, Spec>
+where
+    P: OpProvider,
+    C: ConfigureEvm<Primitives = OpPrimitives> + 'static,
+    Spec: OpHardforks + EthereumHardforks + Send + Sync + 'static,
+{
+    type Block = OpBlock;
+    type ExecutionData = OpExecutionData;
+
+    fn ensure_well_formed_payload(
+        &self,
+        payload: Self::ExecutionData,
+    ) -> Result<RecoveredBlock<Self::Block>, NewPayloadError> {
+        // Use the standalone validation function with chain spec
+        let sealed_block = ensure_well_formed_payload(self.chain_spec.as_ref(), payload)
+            .map_err(|e| NewPayloadError::Other(e.to_string().into()))?;
+
+        // Recover senders for the block
+        let recovered_block = sealed_block
+            .try_recover()
+            .map_err(|_| NewPayloadError::Other("Failed to recover senders".to_string().into()))?;
+
+        Ok(recovered_block)
+    }
+
+    fn validate_block_post_execution_with_hashed_state(
+        &self,
+        _state_updates: &HashedPostState,
+        _block: &RecoveredBlock<Self::Block>,
+    ) -> Result<(), ConsensusError> {
+        // Default implementation - no additional validation needed for Optimism
+        Ok(())
+    }
+}
+
+impl<P, C, Spec> EngineValidator<OpPayloadTypes> for OpPayloadValidator<P, C, Spec>
```

Comment:
it might make sense to split off this trait entirely but we can think about this next